python/cudf/indexing.py

```diff
@@ -0,0 +1,265 @@
+import numbers
+
+import numpy as np
+import pandas as pd
+from numba.cuda.cudadrv.devicearray import DeviceNDArray
+
+import cudf
+from cudf.dataframe.index import Index
+
+
+class _SeriesLocIndexer(object):
+    """
+    Label-based selection
+    """
+
+    def __init__(self, sr):
+        self._sr = sr
+
+    def __getitem__(self, arg):
+        from cudf.dataframe.series import Series
+        if isinstance(arg, (list, np.ndarray, pd.Series, range, Index,
+                            DeviceNDArray)):
+            if len(arg) == 0:
+                arg = Series(np.array([], dtype='int32'))
+            else:
+                arg = Series(arg)
+        if isinstance(arg, Series):
+            if arg.dtype in [np.bool, np.bool_]:
+                return self._sr.iloc[arg]
+            # To do this efficiently we need a solution to
+            # https://github.com/rapidsai/cudf/issues/1087
+            out = Series(
+                [], dtype=self._sr.dtype, index=self._sr.index.__class__([])
+            )
+            for s in arg:
+                out = out.append(self._sr.loc[s:s], ignore_index=False)
+            return out
+        elif _is_single_value(arg):
+            found_index = self._sr.index.find_label_range(arg, None)[0]
+            return self._sr.iloc[found_index]
+        elif isinstance(arg, slice):
+            start_index, stop_index = self._sr.index.find_label_range(
+                arg.start, arg.stop
+            )
+            return self._sr.iloc[start_index:stop_index:arg.step]
+        else:
+            raise NotImplementedError(
+                ".loc not implemented for label type {}".format(
+                    type(arg).__name__
+                )
+            )
+
+
+class _SeriesIlocIndexer(object):
+    """
+    For integer-location based selection.
+    """
+
+    def __init__(self, sr):
+        self._sr = sr
+
+    def __getitem__(self, arg):
+        if isinstance(arg, tuple):
+            arg = list(arg)
+        return self._sr[arg]
```

Comment:
Series[arg] shouldn't really do an iloc, could we update this to hit the underlying integer position gather method?

Reply:
Yes. I'm working on getting `Series[arg]` to do the right thing but that's a non-trivial change and probably better to submit a new PR for that.